### PR TITLE
[fix][kernel] fix timegm month overflow calculation

### LINKED ISSUE #11686: [Bug] 内核mktime函数bug
### RT-Thread Version

5.0.1

### Affected area

Kernel

### Hardware/BSP vendor

Nuvoton

### Architecture

ARM / AArch64

### Board and hardware details

N9H30

### Develop Toolchain

GCC

### Describe the bug

void time_example(void)
{
    struct tm *raw_time;
    struct tm _time;

    _time.tm_year = 126;
    _time.tm_mon = 7;
    _time.tm_mday = 5;
    _time.tm_hour = 8;
    _time.tm_min = 0;
    _time.tm_sec = 0;

    _time.tm_mday += 420;

    time_t timestamp = mktime(&_time);
    raw_time = localtime(&timestamp);
    raw_time->tm_year += 1900;

    //这里应该输出2027/08/29，但实际输出了2027/03/01
    rt_kprintf("%04d/%02d/%02d\n", raw_time->tm_year, raw_time->tm_mon, raw_time->tm_mday);
}

### Other additional context

//临时解决方案
time_t timegm(struct tm * const t)
{
    time_t day;
    time_t i;
    time_t years;

    if(t == RT_NULL)
    {
        rt_set_errno(EFAULT);
        return (time_t)-1;
    }

    years = (time_t)t->tm_year - 70;
    if (t->tm_sec > 60)         /* seconds after the minute - [0, 60] including leap second */
    {
        t->tm_min += t->tm_sec / 60;
        t->tm_sec %= 60;
    }
    if (t->tm_min >= 60)        /* minutes after the hour - [0, 59] */
    {
        t->tm_hour += t->tm_min / 60;
        t->tm_min %= 60;
    }
    if (t->tm_hour >= 24)       /* hours since midnight - [0, 23] */
    {
        t->tm_mday += t->tm_hour / 24;
        t->tm_hour %= 24;
    }
    if (t->tm_mon >= 12)        /* months since January - [0, 11] */
    {
        t->tm_year += t->tm_mon / 12;
        t->tm_mon %= 12;
    }

    while (t->tm_mday > (__spm[t->tm_mon + 1] - __spm[t->tm_mon]))
    {
        int days_in_month = __spm[t->tm_mon + 1] - __spm[t->tm_mon];
        if (t->tm_mon == 1 && __isleap(t->tm_year + 1900))
            days_in_month = 29;
        t->tm_mday -= days_in_month;
        if (++t->tm_mon >= 12) {
            t->tm_mon = 0;
            ++t->tm_year;
        }
    }

    while (t->tm_mday <= 0) {
        if (--t->tm_mon < 0) {
            t->tm_mon = 11;
            --t->tm_year;
        }
        int days_in_month = __spm[t->tm_mon + 1] - __spm[t->tm_mon];
        if (t->tm_mon == 1 && __isleap(t->tm_year + 1900))
            days_in_month = 29;
        t->tm_mday += days_in_month;
    }

    years = (time_t)t->tm_year - 70;

    if (t->tm_year < 70)
    {
        rt_set_errno(EINVAL);
        return (time_t) -1;
    }

    /* Days since 1970 is 365 * number of years + number of leap years since 1970 */
    day = years * 365 + (years + 1) / 4;

    /* After 2100 we have to substract 3 leap years for every 400 years
     This is not intuitive. Most mktime implementations do not support
     dates after 2059, anyway, so we might leave this out for it's
     bloat. */
    if (years >= 131)
    {
        years -= 131;
        years /= 100;
        day -= (years >> 2) * 3 + 1;
        if ((years &= 3) == 3)
            years--;
        day -= years;
    }

    day += t->tm_yday = __spm[t->tm_mon] + t->tm_mday - 1 +
                        (__isleap(t->tm_year + 1900) & (t->tm_mon > 1));

    /* day is now the number of days since 'Jan 1 1970' */
    i = 7;
    t->tm_wday = (int)((day + 4) % i); /* Sunday=0, Monday=1, ..., Saturday=6 */

    i = 24;
    day *= i;
    i = 60;
    return ((day + t->tm_hour) * i + t->tm_min) * i + t->tm_sec;
}
RTM_EXPORT(timegm);

### PR DESCRIPTION
**Pull request by liulangrenaaa**
When normalizing tm_mday, the original code used cumulative day offsets
from __spm[] as if they were the length of the current month. This caused
incorrect dates after crossing month boundaries. For example,
2026/08/05 + 420 days was normalized to 2027/03/01 instead of
2027/09/29.

The existing normalization also did not handle zero or negative tm_mday
values, leaving the broken-down time fields unnormalized.

Use the difference between adjacent __spm[] entries to obtain the actual
month length, with leap-year handling for February. Add backward
normalization for zero and negative days, and normalize negative
tm_sec, tm_min, tm_hour, and tm_mon values consistently with the existing
positive overflow handling.

Add regression tests covering month boundaries, leap years, large day
offsets, zero and negative days, negative time fields, carry boundaries,
combined overflow, and timestamp round-trip consistency.

Close #11686

## 拉取/合并请求描述：(PR description)

为什么提交这份PR (why to submit this PR)

修复 timegm() 在日期字段越界时的归一化错误（Issue #11686）。

原实现使用 __spm[] 的累计天数作为当前月份长度，导致 tm_mday 跨月时计算错误。
例如 2026/08/05 加 420 天，预期结果为 2027/09/29，实际却得到
2027/03/01。

此外，tm_mday 为 0 或负数时也没有向前归一化，可能导致返回时间戳与
struct tm 中的日期字段不一致。

你的解决方案是什么 (what is your solution)

使用 __spm[mon + 1] - __spm[mon] 计算当前月份的实际天数，并正确处理
闰年二月。

增加 tm_mday <= 0 时向前回退月份和年份的归一化逻辑，同时补充
tm_sec、tm_min、tm_hour 和 tm_mon 的负值归一化，使其与已有的正向
越界处理保持一致。

新增 src/utest/time_tc.c 回归测试，覆盖：
- 普通跨月边界
- 闰年二月
- 大天数偏移
- tm_mday 为 0 和负数
- 负 tm_sec/tm_min/tm_hour/tm_mon
- -61、-13 等进位/借位边界
- 正负组合越界
- timestamp 与 UTC round-trip 一致性

请提供验证的bsp和config (provide the config and bsp)

- BSP: bsp/qemu-vexpress-a9
- .config: CONFIG_RT_USING_POSIX_CLOCK=y, CONFIG_RT_UTEST_TIME=y
- action: 本地 QEMU core.time 测试通过

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[clang-format](https://clang.llvm.org/docs/ClangFormat.html) 源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/7.contribution/coding_style_cn.md) This PR has been formatted with clang-format and complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/7.contribution/coding_style_en.md)
- [ ] 如果是新增bsp, 已经添加ci检查到[.github/ALL_BSP_COMPILE.json](https://github.com/RT-Thread/rt-thread/blob/master/.github/ALL_BSP_COMPILE.json)  详细请参考链接[BSP自查](https://www.rt-thread.org/document/site/#/rt-thread-version/rt-thread-standard/development-guide/bsp-selfcheck/bsp_selfcheck)

### CODE REVIEW

**Comment by liulangrenaaa** [comment]
ToolsCI 的失败看起来与本 PR 无关。

当前失败来自最近加入 RT-Thread/packages 的 ai/LiME-TM/Kconfig：

warning: default on the choice symbol LIME_TM_USING_MNIST_FS_EXAMPLE
warning: default on the choice symbol LIME_TM_USING_MNIST_STATIC_EXAMPLE

LiME-TM 是今天由 commit 69de99ddb0b8 加入 packages 的，目前 choice 下的两个 symbol 使用了 default n / default y，kconfiglib 会对此产生 warning。

而 ToolsCI 当前会把任意 Kconfig warning 当成错误并直接失败，因此这里需要先修复 LiME-TM 的 Kconfig，之后再重新运行 ToolsCI。

本 PR 修改的是 timegm() 及对应 utest，与上述 package Kconfig warning 无关。

File: components/libc/compilers/common/ctime.c [hunk 1]
```diff
@@ -422,13 +422,9 @@ time_t timegm(struct tm * const t)
         t->tm_year += t->tm_mon / 12;
         t->tm_mon %= 12;
     }
-    while (t->tm_mday > __spm[1 + t->tm_mon])
+    while (t->tm_mday > __spm[t->tm_mon + 1] - __spm[t->tm_mon] + (__isleap(t->tm_year + 1900) && t->tm_mon == 1))
```
Comment (Rbb666) [inline components/libc/compilers/common/ctime.c]: 新循环只处理“大于当月天数”的情况。对于 2023-01-00、2024-03-00 这类输入，循环会直接跳过，函数虽然返回前一天的时间戳，却仍把结构体保留为非法日期，导致返回时间戳和 struct tm 字段不一致。这不符合 mktime 风格的规范化要求，关联 issue 的讨论也明确提到了负数日期场景。建议增加向前回退的月份/年份规范化逻辑，并补充 tm_mday == 0 和负数日期测试。[关联 issue #11686](https://github.com/RT-Thread/rt-thread/issues/11686)

Comment (liulangrenaaa) [reply components/libc/compilers/common/ctime.c]: ok,  我再check下

Comment (liulangrenaaa) [reply components/libc/compilers/common/ctime.c]: 已更新，补充了 tm_mday <= 0 时向前回退月份/年份的归一化逻辑，并增加了 tm_mday == 0 和负数日期的回归测试。感谢 review。

**Comment by Rbb666** [comment]
有两个问题：

1. 新增用例没有进入 CI 实际执行。RT_UTEST_TIME 是 default n
2. src/utest/time_tc.c 归属不合适。

建议调整为：

```
components/libc/compilers/common/utest/
├── Kconfig
├── SConscript
└── time_tc.c
```

同时：
- 删除本 PR 对 src/utest/Kconfig、src/utest/SConscript 的修改。
- 在 components/utilities/utest/Kconfig 中 rsource 新的 libc 测试 Kconfig。
- 配置名建议改成 RT_UTEST_LIBC_TIME。
- 测试导出名由 core.time 改成 components.libc.time。
- 新增 .github/utest/configs/components/libc.cfg：

```
CONFIG_RT_CONSOLEBUF_SIZE=1024
CONFIG_RT_USING_CI_ACTION=y
CONFIG_RT_USING_POSIX_CLOCK=y
CONFIG_RT_UTEST_LIBC_TIME=y
```

- 在 [utest_auto_run.yml 的 component matrix](https://github.com/RT-Thread/rt-thread/blob/8234a3614e289945cf425e3fcca3979f99a17422/.github/workflows/utest_auto_run.yml#L144-L152) 中增加一个 A9 standard + components/libc.cfg 任务。这样可以明确看到 components.libc.time 被执行

**Comment by liulangrenaaa** [comment]
> 有两个问题：
> 
> 1. 新增用例没有进入 CI 实际执行。RT_UTEST_TIME 是 default n
> 2. src/utest/time_tc.c 归属不合适。
> 
> 建议调整为：
> 
> ```
> components/libc/compilers/common/utest/
> ├── Kconfig
> ├── SConscript
> └── time_tc.c
> ```
> 
> 同时：
> 
> * 删除本 PR 对 src/utest/Kconfig、src/utest/SConscript 的修改。
> * 在 components/utilities/utest/Kconfig 中 rsource 新的 libc 测试 Kconfig。
> * 配置名建议改成 RT_UTEST_LIBC_TIME。
> * 测试导出名由 core.time 改成 components.libc.time。
> * 新增 .github/utest/configs/components/libc.cfg：
> 
> ```
> CONFIG_RT_CONSOLEBUF_SIZE=1024
> CONFIG_RT_USING_CI_ACTION=y
> CONFIG_RT_USING_POSIX_CLOCK=y
> CONFIG_RT_UTEST_LIBC_TIME=y
> ```
> 
> * 在 [utest_auto_run.yml 的 component matrix](https://github.com/RT-Thread/rt-thread/blob/8234a3614e289945cf425e3fcca3979f99a17422/.github/workflows/utest_auto_run.yml#L144-L152) 中增加一个 A9 standard + components/libc.cfg 任务。这样可以明确看到 components.libc.time 被执行

done

**Comment by Rbb666** [comment]
看到测试用例已经运行了，建议在用例中添加点LOG，这样可以直观的看到输出结果

PixPin_2026-08-20_18-35-12

**Comment by liulangrenaaa** [comment]
> 看到测试用例已经运行了，建议在用例中添加点LOG，这样可以直观的看到输出结果
> 
测试日志：
  [I/utest] [----------] [ testcase ] (components.libc.time) started
  [I/utest] large day offset: timestamp=1822204800, normalized=2027-09-29 08:00:00
  [I/utest] zero day: timestamp=1672444800, normalized=2022-12-31 00:00:00
  [I/utest] negative day: timestamp=1709078400, normalized=2024-02-28 00:00:00
  [I/utest] positive overflow: timestamp=1704160921, normalized=2024-01-02 02:02:01
  [I/utest] combined negative overflow: timestamp=1638223079, normalized=2021-11-29 21:57:59
  [I/utest] [  PASSED  ] [ result   ] testcase (components.libc.time)
  [I/utest] [==========] 1 tests from 1 testcase ran.
  [I/utest] [  PASSED  ] [ result   ] 1 tests.

done

**Review by Rbb666 — APPROVED** [review]
